### PR TITLE
libcurl: prevent linker error with zstd

### LINKED ISSUE #3039: [package] libcurl/7.72.0: zstd linker error
the libcurl build system detects libzstd at configure time. if libzstd is detected, it will compile in support for it. if libcurl is compiled as static library and the application does not link with libzstd we get undefined symbols at link time

### Package and Environment Details (include every applicable attribute)
  * Package Name/Version: **libcurl/7.72.0**

### Conan profile (output of `conan profile show default` or `conan profile show <profile>` if custom profile is in use)
```
Configuration for profile default:

[settings]
os=Macos
os_build=Macos
arch=x86_64
arch_build=x86_64
compiler=apple-clang
compiler.version=11.0
compiler.libcxx=libc++
build_type=Release
[options]
[build_requires]
[env]
```

### Steps to reproduce (Include if Applicable)

* build the libcurl recipe locally as static library, when libzstd is in path
* link libcurl into an application

### Logs (Include/Attach if Applicable)
<details><summary>Click to expand log</summary>

```
Undefined symbols for architecture x86_64:
  "_ZSTD_freeDStream", referenced from:
      _zstd_close_writer in libcurl.a(libcurl_la-content_encoding.o)
  "_ZSTD_decompressStream", referenced from:
      _zstd_unencode_write in libcurl.a(libcurl_la-content_encoding.o)
  "_ZSTD_isError", referenced from:
      _zstd_unencode_write in libcurl.a(libcurl_la-content_encoding.o)
  "_ZSTD_createDStream", referenced from:
      _zstd_init_writer in libcurl.a(libcurl_la-content_encoding.o)
ld: symbol(s) not found for architecture x86_64
```

</details>

### PR DESCRIPTION
**Pull request by timblechmann**
autoconf detects zstdlib at configure-time, which can be a system
(locally) installed version. this will lead to linker errors in the
final application

so for now, we disable zstd until the recipe is zstd-aware (by pulling
in the conan package of zstd)

Fixes #3039

Specify library name and version:  **libcurl/7.72**

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [ ] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [ ] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/libcurl/all/conanfile.py [hunk 1]
```diff
@@ -246,6 +246,9 @@ def _get_configure_command_args(self):
 
         params.append("--with-brotli" if self.options.with_brotli else "--without-brotli")
 
+        # TODO: prevent linking of system-installed zstd
+        params.append("--without-zstd")
```
Comment (timblechmann) [inline recipes/libcurl/all/conanfile.py:250]: of course this is a stop-gap workaround until someone finds the time to enable zstd in the recipe properly (zstd was introduces with libcurl 7.72, so the options would have to be done based on the version)

Comment (SpaceIm) [reply recipes/libcurl/all/conanfile.py:250]: Did you check that this option is supported by all libcurl versions supported in CCI?

Comment (madebr) [reply recipes/libcurl/all/conanfile.py:250]: Does this work?
1. Add `with_zstd` to the `options` and `default_options` dict
2. Remove `with_zstd` from `options` in `config_options` if `tools.Version(self.version) < 7.72.0`
3. Add a `_with_zstd` property method that does the following:
    ```python
    @property
    def _with_zstd(self):
        return self.options.get_safe("with_zstd", False)
    ```
4. Add the following argument to the configure script:
`params.append("--with-zstd" if self._with_zstd else "--without-zstd")`

Comment (SpaceIm) [reply recipes/libcurl/all/conanfile.py:250]: I don't think it works, `--with-zstd` doesn't exist in `configure.ac` before `7.72.0`

Comment (madebr) [reply recipes/libcurl/all/conanfile.py:250]: Thanks @SpaceIm , indeed older versions will fail when encountering an unknown option.
Then the steps should be:
1. Add with_zstd to the options and default_options dict
2. Remove with_zstd from options in config_options if tools.Version(self.version) < 7.72.0
3. Add the arguments conditionally to the configure script.
    ```python
    if tools.Version(self.version) >= "7.72.0":
        params.append("--with-zstd" if self.options.with_zstd else "--without-zstd")
    ```

**Comment by madebr** [comment]
@timblechmann 
Please add "Fixes #3039" to your initial post.

File: recipes/libcurl/all/conanfile.py (hunk 1 above)
Comment (SpaceIm) [reply recipes/libcurl/all/conanfile.py:250]: @timblechmann doesn't want to support zstd in this PR (it requires additional work, in `requirements` method, but also in CMake build and `package_info`), just avoid silent link of system zstd during build for autotools builds.

Comment (madebr) [reply recipes/libcurl/all/conanfile.py:250]: Aight! I see!
How about adding a FIXME about missing zstd for versions >= 7.72.0?

**Comment by prince-chrismc** [comment]
> @timblechmann
> Please add "Fixes #3039" to your initial post.

Then we need a new issue to say "add zstd" support :grinning:

**Comment by timblechmann** [comment]
i gave it a try to see how far i get with enabling zstd support in the package. please carefully review, as i might not know 100% what i'm doing.

**Comment by timblechmann** [comment]
btw, i'd love to check the CI for errors, but i seem to have an error 403

File: recipes/libcurl/all/conanfile.py [hunk 2]
```diff
@@ -156,6 +162,10 @@ def requirements(self):
         if self.options.with_brotli:
             self.requires("brotli/1.0.7")
 
+        if tools.Version(self.version) >= "7.72.0":
+            if self.options.with_zstd:
+                self.requires("zstd/1.4.5")
```
Comment (prince-chrismc) [inline recipes/libcurl/all/conanfile.py:167]:
```suggestion
        if self.options.get_safe("with_zstd"):
            self.requires("zstd/1.4.5")
```

File: recipes/libcurl/all/conanfile.py [hunk 3]
```diff
@@ -563,3 +576,7 @@ def package_info(self):
             self.cpp_info.components["curl"].requires.append("zlib::zlib")
         if self.options.with_brotli:
             self.cpp_info.components["curl"].requires.append("brotli::brotli")
+
+        if tools.Version(self.version) >= "7.72.0":
+            if self.options.with_zstd:
+                self.cpp_info.components["curl"].requires.append("zstd::zstd")
```
Comment (prince-chrismc) [inline recipes/libcurl/all/conanfile.py:582]:
```suggestion
        if self.options.get_safe("with_zstd"):
            self.cpp_info.components["curl"].requires.append("zstd::zstd")
```

**Comment by prince-chrismc** [comment]
> btw, i'd love to check the CI for errors, but i seem to have an error 403

It seems like the CI is stuck, usually the CCI bot will post the results

**Comment by uilianries** [comment]
@prince-chrismc Indeed it's a bit unstable, hope infra team can fix it soon. It's Friday, maybe Jenkins has taken a break 🍸

**Comment by prince-chrismc** [comment]
>  It's Friday, maybe Jenkins has taken a break 🍸

❗ without DMing me an invite?!?! I'm offended...

🤣

File: recipes/libcurl/all/conanfile.py [hunk 4]
```diff
@@ -246,6 +255,9 @@ def _get_configure_command_args(self):
 
         params.append("--with-brotli" if self.options.with_brotli else "--without-brotli")
 
+        if tools.Version(self.version) >= "7.72.0":
+            params.append("--with-zstd" if self.options.with_zstd else "--without-zstd")
```
Comment (SpaceIm) [inline recipes/libcurl/all/conanfile.py:259]: What about zstd option in CMake build?

Comment (prince-chrismc) [reply recipes/libcurl/all/conanfile.py:259]: https://github.com/conan-io/conan-center-index/blob/353ac6b1d1c2cd878b6e2903e0460ab54125a75b/recipes/libcurl/all/conanfile.py#L467 should have something added

Comment (madebr) [reply recipes/libcurl/all/conanfile.py:259]: https://github.com/curl/curl/blob/e6b21d422e631a7c0cc81abf956af179b3b4c5e8/CMakeLists.txt#L661-L675
A `CURL_ZSTD` definition.

The name of the zstd cmake find_package is custom.
Zstd vs 
https://github.com/conan-io/conan-center-index/blob/4ac63884d528d6abb2d4740acb88d5cbd3d6b76a/recipes/zstd/all/conanfile.py#L70

Comment (SpaceIm) [reply recipes/libcurl/all/conanfile.py:259]: Yes, zstd only exports its targets since 1.4.5, so it makes sense for libcurl to use a custom module file.

**Pull request closed without merging**